src/parallel/communicator.C

```diff
@@ -207,30 +223,30 @@ void Communicator::nonblocking_barrier (Request & req) const
     }
 }
 #else
-void Communicator::barrier () const {}
+void Communicator::nonblocking_barrier (Request & /*req*/) const {}
 #endif
 
 
-MessageTag Communicator::get_unique_tag(int tagvalue) const
+MessageTag Communicator::get_unique_tag(int /*tagvalue*/) const
 {
-  if (used_tag_values.count(tagvalue))
-    {
-      // Get the largest value in the used values, and pick one
-      // larger
-      tagvalue = used_tag_values.rbegin()->first+1;
-      libmesh_assert(!used_tag_values.count(tagvalue));
-    }
-  used_tag_values[tagvalue] = 1;
+  // Don't give out tag values greater than what MPI can handle!
+  if (_next_tag == _max_tag)
+    _next_tag = 0;
+
+  auto new_tag = _next_tag++;
+
+  // We don't want to simply get the largest one
+  // because it might be near MPI_TAG_UB and then
+  // we can end up in a sticky situation
+  // Most-likely we will amost never hit a
+  // tag that is in use because we have
+  // ~2 billion of them
+  while(used_tag_values.count(new_tag))
+    new_tag++;
 
-  // #ifndef NDEBUG
-  //   // Make sure everyone called get_unique_tag and make sure
-  //   // everyone got the same value
-  //   int maxval = tagvalue;
-  //   this->max(maxval);
-  //   libmesh_assert_equal_to (tagvalue, maxval);
-  // #endif
+  used_tag_values[new_tag] = 1;
 
```

Comment:
I dug into SVN history to see why this was commented out, and apparently Ben had a use case in which it was expected to fail, because tags were being grabbed on a subset of processors.  So I feel a little better about going to the effort of respecting tagvalue where it's used while avoiding using it where possible.

Actually, that two-part system ought to make sanity checking easier.  Where the user requests an automatic tag value, we can be sure they want to keep _next_tag in sync, and we can restore this check.